zenith_utils/src/postgres_backend.rs

```diff
@@ -0,0 +1,180 @@
+use anyhow::Result;
+use bytes::{Bytes, BytesMut};
+use log::*;
+use std::io;
+use std::io::{BufReader, Write};
+use std::net::{Shutdown, TcpStream};
+
+pub trait Handler {
+    /// Handle single query.
+    /// postgres_backend will issue ReadyForQuery after calling this (this
+    /// might be not what we want after CopyData streaming, but currently we don't
+    /// care).
+    fn process_query(&mut self, pgb: &mut PostgresBackend, query_string: Bytes) -> Result<()>;
+    /// Called on startup packet receival, allows to process params.
+    fn startup(&mut self, _pgb: &mut PostgresBackend, _sm: &FeStartupMessage) -> Result<()> {
+        Ok(())
+    }
+}
+
+pub struct PostgresBackend {
+    // replication.rs wants to handle reading on its own in separate thread, so
+    // wrap in Option to be able to take and transfer the BufReader. Ugly, but I
+    // have no better ideas.
+    stream_in: Option<BufReader<TcpStream>>,
+    stream_out: TcpStream,
+    // Output buffer. c.f. BeMessage::write why we are using BytesMut here.
+    buf_out: BytesMut,
+    init_done: bool,
+}
+
+// In replication.rs a separate thread is reading keepalives from the
+// socket. When main one finishes, tell it to get down by shutdowning the
+// socket.
+impl Drop for PostgresBackend {
+    fn drop(&mut self) {
+        let _res = self.stream_out.shutdown(Shutdown::Both);
+    }
+}
+
+impl PostgresBackend {
+    pub fn new(socket: TcpStream) -> Result<Self, std::io::Error> {
+        let mut pb = PostgresBackend {
+            stream_in: None,
+            stream_out: socket,
+            buf_out: BytesMut::with_capacity(10 * 1024),
+            init_done: false,
+        };
+        // if socket cloning fails, report the error and bail out
+        pb.stream_in = match pb.stream_out.try_clone() {
+            Ok(read_sock) => Some(BufReader::new(read_sock)),
+            Err(error) => {
+                let errmsg = format!("{}", error);
+                let _res = pb.write_message_noflush(&BeMessage::ErrorResponse(errmsg));
+                return Err(error);
+            }
+        };
+        Ok(pb)
+    }
+
+    /// Get direct reference (into the Option) to the read stream.
+    fn get_stream_in(&mut self) -> Result<&mut BufReader<TcpStream>> {
+        match self.stream_in {
+            Some(ref mut stream_in) => Ok(stream_in),
+            None => bail!("stream_in was taken"),
+        }
+    }
+
+    pub fn take_stream_in(&mut self) -> Option<BufReader<TcpStream>> {
+        self.stream_in.take()
+    }
+
+    /// Read full message or return None if connection is closed.
+    pub fn read_message(&mut self) -> Result<Option<FeMessage>> {
+        if !self.init_done {
+            FeStartupMessage::read(self.get_stream_in()?)
+        } else {
+            FeMessage::read(self.get_stream_in()?)
+        }
+    }
+
+    /// Write message into internal output buffer.
+    pub fn write_message_noflush(&mut self, message: &BeMessage) -> io::Result<()> {
```

Comment:
BTW we could replace `Result<()>` with `Result<&mut Self>` to support method chaining, i.e.

```rust
backend
  .write_message_noflush(msg1)?
  .write_message_noflush(msg2)?
  .write_message_noflush(msg3)?
  .fush()?;
```

Reply:
Same goes for the rest of the methods.

Reply:
Nice idea, done that.